text/0000-import-trait-methods.md

```diff
@@ -0,0 +1,238 @@
+
+```rust
+use Default::default;
+
+struct S {
+    a: HashMap<i32, i32>,
+}
+
+impl S {
+    fn new() -> S {
+        S {
+            a: default()
+        }
+    }
+}
+```
+desugars to 
+```rust
+struct S {
+    a: HashMap<i32, i32>,
+}
+
+impl S {
+    fn new() -> S {
+        S {
+            a: Default::default()
+        }
+    }
+}
+```
+And a call
+```rust
+use Trait::method as m;
+m(x, y, z);
+```
+desugars to
+```rust
+Trait::method(x, y, z);
+```
+
+Additionally, the syntax
+```rust
+use some_module::Trait::self;
+```
+is sugar for
+```rust
+use some_module::Trait;
+```
+
+Finally, given traits 
+```rust
+trait Super {
+    fn f();
+}
+
+trait Sub : Super {
+}
+```
+the usage
+```rust
+use module::Sub::f;
+f();
+```
+desugars to
+```rust
+use module::Sub::f;
+Super::f();
+```
+**not** `Sub::f();` as that desugaring will cause compiler errors, see https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=51bef9ba69ce1fc20248e987bf106bd4.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Calls to `default` are less explicit than calls to `Default::default`, likewise for any other trait.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Why is this design the best in the space of possible designs?
+
+This design is minimalist, it adds no extra syntax, instead providing a natural extension of existing syntax to support a feature that is frequently requested. Users might very well already expect this feature, with this exact syntax, to be present in Rust, and surprised when it isn't.
+
+## What other designs have been considered and what is the rationale for not choosing them?
+
+In [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Writing.20an.20RFC.20for.20.60use.20Default.3A.3Adefault.60/near/427795694), there was some discussion of whether `use Trait::method` should bring `Trait` into scope. There are three possibilities:
+
+1. It does not - this may be unexpected, but maybe not
+2. It does - then `value.other_method_from_the_same_trait()` will work as well, this may be unexpected too
+3. It does, but only for method, that's something new for the language (need new more fine-grained tracking of traits in scope)
+
+Option 1 is what is proposed here. It has the simplest semantics, and I believe it best matches the user intent when they import a trait method; the desire is to make that method available as-if it were a regular function. Furthermore, it is more minimalist than the other two options in the sense that you can get to option 2 simply by importing the trait also. Option 3 seems like extra complexity for almost no added value.
+
+An earlier version of this RFC proposed not allowing `use Trait::super_trait_method`. This was changed because comments indicated this usecase would be common.
+
+## What is the impact of not doing this?
+
+Users of the language continue to create helper methods to access trait methods with infix syntax.
```

Comment:
I think people writing `defualt()` instead of `T::default()` is a _major_ drawback here, because it's much less clear.

Reply:
> This is false.

This is true; users of the language already do this, see my example from Bevy in the motivation section. To make it clear, this feature merely simplifies something that is already possible on stable rust today. For any trait method `Trait::method`, you can (and people do!) get this feature using the following:
```rust
fn method<T: Trait>(insert_method_sig_here) -> insert_method_return_here {
    T::method(args)
}
```
A more reasonable argument is that we shouldn't make it easier to do this, but the fact that people do it is undebatable.

Reply:
I've never seen someone make a free method for this, but regardless..

If `T` is a type then `T::default()` is more common than `Default::default()` because it's unambiguous and unclear where `Default::default()` is preferable, excapt maybe in macros.  

My point is both this fn and this RFC are worse style.

Reply:
I disagree with your premise about style (and the fitness of your suggested alternative), and I have updated the drawbacks section with my thoughts.